File: whitespace/core/src/main/java/dev/whitespace/core/service/bundler/domain/Html.java
package dev.whitespace.core.service.bundler.domain;


import org.jspecify.annotations.Nullable;
import org.w3c.dom.Attr;

import java.util.ArrayList;
import java.util.List;

public class Html {
    public static class Node {
        public Integer nodeType = -1;
        public @Nullable String tagName;
        public @Nullable String nodeValue;
        public @Nullable String nodeName;
        public @Nullable String namespace;
        public @Nullable List<Attribute> attributes;
        public @Nullable List<Property> properties;
        public @Nullable List<Node> children;
        public @Nullable Node shadowRoot;

		public boolean hasTagName(String tagName) {
			if (this.tagName == null) {
				return false;
			}
			return this.tagName.equalsIgnoreCase(tagName);
		}

		public @Nullable Attribute getAttribute(String attributeName) {
			for (var attribute : attributes) {
				if (attribute.name.equalsIgnoreCase(attributeName)) {
					return attribute;
				}
			}
			return null;
		}

		public @Nullable Attribute getAttributeNS(String attributeName, String namespace) {
			for (var attribute : attributes) {
				if (attribute.namespace == null) {
					continue;
				}

				if (attribute.name.equalsIgnoreCase(attributeName) && attribute.namespace.equals(namespace)) {
					return attribute;
				}
			}
			return null;
		}

		public List<Attribute> getAttributeStartingWith(String prefix) {
			return attributes.stream().filter(attr -> attr.name.startsWith(prefix)).toList();
		}

		public static class Attribute {
            public String name;
            public String value;
            public @Nullable String namespace;
        }

        public static class Property {
            public String name;
            public String value;
        }
    }
}

File: whitespace/src/main/kotlin/dev/whitespace/domain/recording/Html.kt
package dev.whitespace.domain.recording

import dev.whitespace.domain.recording.serializer.NullAsEmptyListSerializer
import kotlinx.serialization.Serializable

@Serializable
data class HtmlNode(
    val nodeType: Int,
    val tagName: String?,
    val nodeValue: String?,
    val nodeName: String?,
    val namespace: String?,
    val attributes: List<HtmlNodeAttribute>?,
    val properties: List<HtmlNodeProperty>?,
    val children: List<HtmlNode>?,
    val shadowRoot: HtmlNode?,
) {
    fun hasTagName(tagName: String): Boolean {
        return when (this.tagName) {
            is String -> this.tagName.equals(tagName, ignoreCase = true)
            null -> false
        }
    }

    fun getAttribute(name: String): HtmlNodeAttribute? {
        return attributes?.firstOrNull { it.name.equals(name, ignoreCase = true) }
    }

    fun getAttributeNs(name: String, ns: String): HtmlNodeAttribute? {
        return attributes?.firstOrNull {
            it.namespace != null &&
            it.namespace.equals(ns, ignoreCase = true) &&
            it.name.equals(name, ignoreCase = true)
        }
    }
}

//func getAttributeNS(node *recording.HTMLNode, name string, namespace string) *recording.HTMLNodeAttribute {
//    for i := range node.Attributes {
//        attr := &node.Attributes[i]
//        if attr.Namespace == nil {
//            continue
//        }
//
//        if strings.EqualFold(attr.Name, name) && strings.EqualFold(*attr.Namespace, namespace) {
//            return attr
//        }
//    }
//    return nil
//}


@Serializable
data class HtmlNodeAttribute(
    val name: String,
    val value: String,
    val namespace: String?,
)

@Serializable
data class HtmlNodeProperty(
    val name: String,
    val value: DomArgument,
)